staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go

```diff
@@ -51,3 +58,140 @@ func Required[T comparable](fldPath *field.Path, value T) field.ErrorList {
 	}
 	return nil
 }
+
+// Union verifies that exactly one member of a union is specified.
+//
+// UnionMembership must define all the members of the union.
+//
+// For example:
+//
+//	var abcUnionMembership := schema.NewUnionMembership("a", "b", "b")
+//	func ValidateABC(fldPath, *field.Path, in *ABC) (errs fields.ErrorList) {
+//		errs = append(errs, Union(fldPath, in, abcUnionMembership, in.A, in.B, in.C)...)
+//		return errs
+//	}
+func Union(fldPath *field.Path, in any, union *UnionMembership, fieldValues ...any) field.ErrorList {
+	if len(union.members) != len(fieldValues) {
+		return field.ErrorList{field.InternalError(fldPath, fmt.Errorf("unexpected difference in length between fields defined in UnionMembership as fieldValues"))}
+	}
+	var specifiedMember *string
+	for i, fieldValue := range fieldValues {
+		rv := reflect.ValueOf(fieldValue)
+		if rv.IsValid() && !rv.IsZero() {
+			m := union.members[i]
+			if specifiedMember != nil && *specifiedMember != m.memberName {
+				return field.ErrorList{field.Invalid(fldPath, "", fmt.Sprintf("must not specify %s, must specify one of: %s", union.specifiedFields(fieldValues), union.allFields()))}
```

Comment:
As we reach completion of this, I'd like to think hard about the most useful phrasing of the error message.

Reply:
Please check me on this.  I've made an effort to produce errors that focus on what fields should be specified, e.g.:

`must not specify "b" and "d", must specify one of: "a", "b", "c" or "d"` (non-discriminated)

`"b" must not be specified when "type" is "C"` (discriminated)

The major alternative I considered was to include the concept of "union" directly in the error messages. Happy to consider phrasing taking that approach if it helps with clarity.